### DISCUSSION TITLE
Fanatac HID Input Report

### DISCUSSION THREAD

**Discussion by OOOllie** [gotzl/hid-fanatecff#84] [Q&A] [answered]
I was wondering if you guys had any resources on learning about the structure of the input report and how I would control it for force feedback. Im implementing support for a Fanatec steering wheel via WebHID and was looking for some resources!

**Comment by gotzl** [comment]
Hmm.. There is my poor-mans-testing-script https://github.com/gotzl/hid-fanatecff-tools/blob/master/tools/hid-fanatecff-test.py
Maybe this helps? It doesn't use HID reports, though.. Never heard of WebHID ;)

That said, shouldn't the API allow you to read out the report? Then sending cmds should be very much similar to what's in the C code ...
Also just found https://developer.chrome.com/docs/capabilities/web-apis/hid-examples with example for PS4 controller, should contain everythinkg you need, but might be a bit overwhelming ^^

And this 'minimal' example pretty much worked also straight away: https://github.com/WICG/webhid/blob/main/EXPLAINER.md#example (had to add a `device = devices[0]`) and logs the inputs received from the wheel.

In any case, the FFB stuff you'll have to do yourself, there is no access from the WebHID to linux FFB API. So, you've to basically implement what's in the hid-fanatec C code ...

**Reply by OOOllie** [reply]
Thanks for getting back really quickly! Definitely a lot for me to read and useful resources! WebHID is just HID devices through your browser. I've already got data out of the wheel to control a car! 

I just want to be able to send the force feedback from browser to the wheel! So was looking to see if anyone had the HID report to the device to send from the browser and what to set! I'll have a read later today and see if I can work it out from what you sent!

**Reply by OOOllie** [reply]
So I've done some further reading and it seems very feasible to do the FFB over WebHID. When you send an output report you are essentially sending those commands you shared via the python test script with the report id being the base command I believe. Im going to test it out tomorrow and see where I get with it!

**Comment by OOOllie** [comment]
@gotzl 

@gotzl So I've done some research and I've managed to get the HID report descriptor for the device:

0x05, 0x01,        // Usage Page (Generic Desktop Ctrls)
0x09, 0x04,        // Usage (Joystick)
0xA1, 0x01,        // Collection (Application)
0x85, 0x01,        //   Report ID (1)
0x09, 0x39,        //   Usage (Hat switch)
0x15, 0x00,        //   Logical Minimum (0)
0x25, 0x07,        //   Logical Maximum (7)
0x35, 0x00,        //   Physical Minimum (0)
0x46, 0x3B, 0x01,  //   Physical Maximum (315)
0x65, 0x14,        //   Unit (System: English Rotation, Length: Centimeter)
0x75, 0x04,        //   Report Size (4)
0x95, 0x01,        //   Report Count (1)
0x81, 0x02,        //   Input (Data,Var,Abs,No Wrap,Linear,Preferred State,No Null Position)
0x65, 0x00,        //   Unit (None)
0x25, 0x01,        //   Logical Maximum (1)
0x45, 0x01,        //   Physical Maximum (1)
0x05, 0x09,        //   Usage Page (Button)
0x19, 0x01,        //   Usage Minimum (0x01)
0x29, 0x6C,        //   Usage Maximum (0x6C)
0x75, 0x01,        //   Report Size (1)
0x95, 0x7C,        //   Report Count (124)
0x81, 0x02,        //   Input (Data,Var,Abs,No Wrap,Linear,Preferred State,No Null Position)
0x05, 0x01,        //   Usage Page (Generic Desktop Ctrls)
0x09, 0x30,        //   Usage (X)
0x09, 0x32,        //   Usage (Z)
0x09, 0x35,        //   Usage (Rz)
0x09, 0x31,        //   Usage (Y)
0x27, 0xFF, 0xFF, 0x00, 0x00,  //   Logical Maximum (65534)
0x47, 0xFF, 0xFF, 0x00, 0x00,  //   Physical Maximum (65534)
0x75, 0x10,        //   Report Size (16)
0x95, 0x04,        //   Report Count (4)
0x81, 0x02,        //   Input (Data,Var,Abs,No Wrap,Linear,Preferred State,No Null Position)
0x09, 0x33,        //   Usage (Rx)
0x09, 0x34,        //   Usage (Ry)
0x15, 0x80,        //   Logical Minimum (-128)
0x25, 0x7F,        //   Logical Maximum (127)
0x35, 0x80,        //   Physical Minimum (-128)
0x45, 0x7F,        //   Physical Maximum (127)
0x75, 0x08,        //   Report Size (8)
0x95, 0x02,        //   Report Count (2)
0x81, 0x02,        //   Input (Data,Var,Abs,No Wrap,Linear,Preferred State,No Null Position)
0x09, 0x36,        //   Usage (Slider)
0x15, 0x00,        //   Logical Minimum (0)
0x26, 0xFF, 0x00,  //   Logical Maximum (255)
0x35, 0x00,        //   Physical Minimum (0)
0x46, 0xFF, 0x00,  //   Physical Maximum (255)
0x95, 0x01,        //   Report Count (1)
0x81, 0x02,        //   Input (Data,Var,Abs,No Wrap,Linear,Preferred State,No Null Position)
0x09, 0x37,        //   Usage (Dial)
0x15, 0x80,        //   Logical Minimum (-128)
0x25, 0x7F,        //   Logical Maximum (127)
0x35, 0x80,        //   Physical Minimum (-128)
0x45, 0x7F,        //   Physical Maximum (127)
0x95, 0x01,        //   Report Count (1)
0x81, 0x02,        //   Input (Data,Var,Abs,No Wrap,Linear,Preferred State,No Null Position)
0x06, 0x00, 0xFF,  //   Usage Page (Vendor Defined 0xFF00)
0x09, 0x01,        //   Usage (0x01)
0x95, 0x05,        //   Report Count (5)
0x81, 0x02,        //   Input (Data,Var,Abs,No Wrap,Linear,Preferred State,No Null Position)
0x09, 0x02,        //   Usage (0x02)
0x95, 0x07,        //   Report Count (7)
0x91, 0x02,        //   Output (Data,Var,Abs,No Wrap,Linear,Preferred State,No Null Position,Non-volatile)
0xC0,              // End Collection
0x05, 0x01,        // Usage Page (Generic Desktop Ctrls)
0x09, 0x04,        // Usage (Joystick)
0xA1, 0x01,        // Collection (Application)
0x85, 0x02,        //   Report ID (2)
0x09, 0x39,        //   Usage (Hat switch)
0x09, 0x39,        //   Usage (Hat switch)
0x09, 0x39,        //   Usage (Hat switch)
0x09, 0x39,        //   Usage (Hat switch)
0x15, 0x00,        //   Logical Minimum (0)
0x25, 0x07,        //   Logical Maximum (7)
0x35, 0x00,        //   Physical Minimum (0)
0x46, 0x3B, 0x01,  //   Physical Maximum (315)
0x65, 0x14,        //   Unit (System: English Rotation, Length: Centimeter)
0x75, 0x04,        //   Report Size (4)
0x95, 0x04,        //   Report Count (4)
0x81, 0x02,        //   Input (Data,Var,Abs,No Wrap,Linear,Preferred State,No Null Position)
0x65, 0x00,        //   Unit (None)
0x25, 0x01,        //   Logical Maximum (1)
0x45, 0x01,        //   Physical Maximum (1)
0x05, 0x09,        //   Usage Page (Button)
0x19, 0x01,        //   Usage Minimum (0x01)
0x29, 0x3F,        //   Usage Maximum (0x3F)
0x75, 0x01,        //   Report Size (1)
0x95, 0x40,        //   Report Count (64)
0x81, 0x02,        //   Input (Data,Var,Abs,No Wrap,Linear,Preferred State,No Null Position)
0x05, 0x01,        //   Usage Page (Generic Desktop Ctrls)
0x09, 0x30,        //   Usage (X)
0x09, 0x31,        //   Usage (Y)
0x09, 0x32,        //   Usage (Z)
0x09, 0x33,        //   Usage (Rx)
0x09, 0x34,        //   Usage (Ry)
0x09, 0x35,        //   Usage (Rz)
0x09, 0x36,        //   Usage (Slider)
0x09, 0x36,        //   Usage (Slider)
0x27, 0xFF, 0xFF, 0x00, 0x00,  //   Logical Maximum (65534)
0x47, 0xFF, 0xFF, 0x00, 0x00,  //   Physical Maximum (65534)
0x75, 0x10,        //   Report Size (16)
0x95, 0x08,        //   Report Count (8)
0x81, 0x02,        //   Input (Data,Var,Abs,No Wrap,Linear,Preferred State,No Null Position)
0x09, 0x37,        //   Usage (Dial)
0x09, 0x37,        //   Usage (Dial)
0x09, 0x37,        //   Usage (Dial)
0x09, 0x37,        //   Usage (Dial)
0x15, 0x80,        //   Logical Minimum (-128)
0x25, 0x7F,        //   Logical Maximum (127)
0x35, 0x80,        //   Physical Minimum (-128)
0x45, 0x7F,        //   Physical Maximum (127)
0x75, 0x08,        //   Report Size (8)
0x95, 0x04,        //   Report Count (4)
0x81, 0x02,        //   Input (Data,Var,Abs,No Wrap,Linear,Preferred State,No Null Position)
0x06, 0x00, 0xFF,  //   Usage Page (Vendor Defined 0xFF00)
0x09, 0x01,        //   Usage (0x01)
0x95, 0x03,        //   Report Count (3)
0x81, 0x02,        //   Input (Data,Var,Abs,No Wrap,Linear,Preferred State,No Null Position)
0x09, 0x02,        //   Usage (0x02)
0x95, 0x07,        //   Report Count (7)
0x91, 0x02,        //   Output (Data,Var,Abs,No Wrap,Linear,Preferred State,No Null Position,Non-volatile)
0xC0,              // End Collection
0x05, 0x01,        // Usage Page (Generic Desktop Ctrls)
0x09, 0x3A,        // Usage (Counted Buffer)
0xA1, 0x01,        // Collection (Application)
0x85, 0xFF,        //   Report ID (-1)
0x05, 0x01,        //   Usage Page (Generic Desktop Ctrls)
0x09, 0x3B,        //   Usage (Byte Count)
0x15, 0x00,        //   Logical Minimum (0)
0x26, 0xFF, 0x00,  //   Logical Maximum (255)
0x35, 0x00,        //   Physical Minimum (0)
0x46, 0xFF, 0x00,  //   Physical Maximum (255)
0x75, 0x08,        //   Report Size (8)
0x95, 0x3F,        //   Report Count (63)
0x81, 0x02,        //   Input (Data,Var,Abs,No Wrap,Linear,Preferred State,No Null Position)
0x06, 0x00, 0xFF,  //   Usage Page (Vendor Defined 0xFF00)
0x09, 0x01,        //   Usage (0x01)
0x95, 0x3F,        //   Report Count (63)
0x91, 0x02,        //   Output (Data,Var,Abs,No Wrap,Linear,Preferred State,No Null Position,Non-volatile)
0xC0,              // End Collection

// 296 bytes

So ive managed to get the HID report descriptor of the device that defines the output to the device over HID but unsure on how this would colerate to the commands any advice?

**Reply by gotzl** [reply]
Well, the descriptor describes the fields inside the data, you've to check the HID spec for what all these things mean ;) As an example, `Usage Page (Button)` + `Report Size (1)` + `Report Count (124)` says, that these 124bytes correspond to buttons... 
In my case, the kernel does this decoding already, so I didn't have to care...

How to extract the interesting bytes from the data, maybe check the PS4 controller example: https://github.com/TheBITLINK/WebHID-DS4/blob/f43a34efff7dd2f210375944e0c9b9309ea354d6/src/index.ts#L122

**Comment by OOOllie** [comment, marked as answer]
I've actually just solved by using Wireshark with your demo I was able to construct the code to do it:
device.sendReport(0x01, new Uint8Array([1, 8, 128]))
In your case, your CMD_BASE is actually just the HID report Id

Would be really nice if you could compile a list of commands and maintain it for me but will mark this as answered! Just want to say thanks so much for helping you've saved me a bunch of time!
